packages/config-loader/src/schema/collect.ts

```diff
@@ -183,17 +184,40 @@ async function compileTsSchemas(paths: string[]) {
   const tsSchemas = paths.map(path => {
     let value;
     try {
-      value = generateSchema(
+      const generator = buildGenerator(
         program,
-        // All schemas should export a `Config` symbol
-        'Config',
         // This enables the use of these tags in TSDoc comments
         {
           required: true,
           validationKeywords: ['visibility', 'deepVisibility', 'deprecated'],
         },
         [path.split(sep).join('/')], // Unix paths are expected for all OSes here
-      ) as JsonObject | null;
+      );
+
+      // All schemas should export a `Config` symbol
+      value = generator?.getSchemaForSymbol('Config') as JsonObject | null;
+
+      // This makes sure that no additional symbols are defined in the schema. We don't allow
+      // this because they share a global namespace and will be merged together, leading to
+      // unpredictable behavior.
+      const userSymbols = new Set(generator?.getUserSymbols());
+      userSymbols.delete('Config');
+      if (userSymbols.size !== 0) {
+        const names = Array.from(userSymbols).join("', '");
+        throw new Error(
+          `Invalid configuration schema in ${path}, additional symbol definitions are not allowed, found '${names}'`,
+        );
+      }
+
+      // This makes sure that no unsupported types are used in the schema, for example `Record<,>`.
+      // The generator will extract these as a schema reference, which will in turn be broken for our usage.
+      const reffedDefs = Object.keys(generator?.ReffedDefinitions ?? {});
+      if (reffedDefs.length !== 0) {
+        const lines = reffedDefs.join(`${EOL}  `);
+        throw new Error(
+          `Invalid configuration schema in ${path}, the following definitions are not supported:${EOL}${EOL}  ${lines}`,
```

Comment:
noted double EOL var; was that intentional?

Reply:
Yep! To give a bit more spacing between the message itself and the horrible blobs of JSON :p